### ISSUE TITLE
Systemctl not working: (code=dumped, signal=SEGV) 

### ISSUE THREAD

**Issue by gabridome** [shesek/spark-wallet#27]
spark-wallet --version
0.1.1
OS: 4.15.0-34-generic #37-Ubuntu SMP
My spark-wallet.system file:
````
[Unit]
Description=Spark Lightning Wallet
Requires=network.target
After=network.target

[Service]
User=bitcoin
Group=bitcoin
#Restart=on-failure

ExecStart=/home/bitcoin/.npm-global/bin/spark-wallet --no-test-conn
# --no-test-conn because c-lightning might not be ready when we boot up.
# this delays opening the connection until we get the first API call from a user.
# if c-lightning is running as a systemd service, you should make this service depend on it instead.

SyslogIdentifier=spark-wallet
PIDFile=/var/run/spark-wallet.pid
StandardInput=null
StandardOutput=syslog
StandardError=syslog

# Hardening measures
PrivateTmp=true
ProtectSystem=full
NoNewPrivileges=true
PrivateDevices=true
MemoryDenyWriteExecute=true

[Install]
WantedBy=multi-user.target

````
my commands and outputs:

````
bitcoin@bitseed3:~$ sudo systemctl daemon-reload                                                                   │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:19: Unknown lvalue 'RuntimeDirectory' i
bitcoin@bitseed3:~$ sudo systemctl start spark-wallet.service                                                      │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:21: Unknown lvalue 'User' in section 'U
bitcoin@bitseed3:~$ sudo systemctl status spark-wallet.service                                                     │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:22: Unknown lvalue 'Group' in section '
● spark-wallet.service - Spark Lightning Wallet                                                                    │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:23: Unknown lvalue 'Type' in section 'U
   Loaded: loaded (/etc/systemd/system/spark-wallet.service; enabled; vendor preset: enabled)                      │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:24: Unknown lvalue 'PIDFile' in section
   Active: failed (Result: core-dump) since Fri 2018-09-28 17:18:08 CEST; 5s ago                                   │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:25: Unknown lvalue 'Restart' in section
  Process: 10544 ExecStart=/home/bitcoin/.npm-global/bin/spark-wallet --no-test-conn (code=dumped, signal=SEGV)    │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:31: Unknown lvalue 'PrivateTmp' in sect
 Main PID: 10544 (code=dumped, signal=SEGV)                                                                        │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:34: Unknown lvalue 'ProtectSystem' in s
                                                                                                                   │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:38: Unknown lvalue 'NoNewPrivileges' in
Sep 28 17:18:08 bitseed3 systemd[1]: Started Spark Lightning Wallet.                                               │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:42: Unknown lvalue 'PrivateDevices' in 
Sep 28 17:18:08 bitseed3 systemd[1]: spark-wallet.service: Main process exited, code=dumped, status=11/SEGV        │Sep 28 17:11:17 bitseed3 systemd[1]: /etc/systemd/system/lightningd.service:45: Unknown lvalue 'MemoryDenyWriteExec
Sep 28 17:18:08 bitseed3 systemd[1]: spark-wallet.service: Failed with result 'core-dump'.                         │Sep 28 17:11:17 bitseed3 lnd[2142]: 2018-09-28 17:11:17.551 [DBG] PEER: Received QueryShortChanIDs(chain_hash=00000
````
I found a coredump in the logs I think is related
````
systemd-coredump[11355]: Process 11335 (node) of user 1000 dumped core.
                                                  
                                                  Stack trace of thread 11335:
                                                  #0  0x0000000000fe8280 _ZN2v88internal18RegExpResultsCache5ClearEPNS0_10FixedArrayE (node)
                                                  #1  0x0000000000dee26a _ZN2v88internal4Heap19MarkCompactPrologueEv (node)
                                                  #2  0x0000000000dee92f _ZN2v88internal4Heap11MarkCompactEv (node)
                                                  #3  0x0000000000e0a119 _ZN2v88internal4Heap24PerformGarbageCollectionENS0_16GarbageCollectorENS_15GCCallbackFlagsE (node)
                                                  #4  0x0000000000e0afe8 _ZN2v88internal4Heap14CollectGarbageENS0_15AllocationSpaceENS0_23GarbageCollectionReasonENS_15GCCallbackFlagsE.constprop.934 (node)
                                                  #5  0x0000000000e0bb6c _ZN2v88internal4Heap12ReserveSpaceEPSt6vectorINS1_5ChunkESaIS3_EEPS2_IPhSaIS7_EE (node)
                                                  #6  0x00000000010634b7 _ZN2v88internal12Deserializer12ReserveSpaceEv (node)
                                                  #7  0x0000000001072a7e _ZN2v88internal19StartupDeserializer15DeserializeIntoEPNS0_7IsolateE (node)
                                                  #8  0x0000000000eb14f6 _ZN2v88internal7Isolate4InitEPNS0_19StartupDeserializerE (node)
                                                  #9  0x0000000001072141 _ZN2v88internal8Snapshot10InitializeEPNS0_7IsolateE (node)
                                                  #10 0x0000000000a87fc8 _ZN2v814IsolateNewImplEPNS_8internal7IsolateERKNS_7Isolate12CreateParamsE (node)
                                                  #11 0x00000000008d51ad _ZN4node5StartEP9uv_loop_siPKPKciS5_ (node)
                                                  #12 0x00000000008d4900 _ZN4node5StartEiPPc (node)
                                                  #13 0x00007f2ffdee1b97 __libc_start_main (libc.so.6)
                                                  #14 0x000000000089e0e1 _start (node)
````

**Comment by vswee** [comment]
Not likely a bug.
This occurs when npm is installed with root privileges. 
Check or fix your npm installation first. Then try again.
Good luck.

**Comment by gabridome** [comment]
Thank you. I will try.

**Comment by gabridome** [comment]
Now the system seems started but `journalctl -xe`says:
````
● spark-wallet.service - Spark Lightning Wallet
   Loaded: loaded (/etc/systemd/system/spark-wallet.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/spark-wallet.service.d
           └─override.conf
   Active: active (running) since Thu 2018-11-01 11:13:56 UTC; 396ms ago
 Main PID: 1404 (node)
    Tasks: 6 (limit: 4520)
   CGroup: /system.slice/spark-wallet.service
           └─1404 node /usr/bin/spark-wallet --onion

Nov 01 11:13:56 bitseed3 systemd[1]: Started Spark Lightning Wallet.
Nov 01 11:13:57 bitseed3 spark-wallet[1404]: #
Nov 01 11:13:57 bitseed3 spark-wallet[1404]: # Fatal error in , line 0
Nov 01 11:13:57 bitseed3 spark-wallet[1404]: # Check failed: SetPermissions(protect_start, protect_size, PageAllocator::kReadExecute).
Nov 01 11:13:57 bitseed3 spark-wallet[1404]: #
````

**Comment by shesek** [comment]
Are you able to start Spark with `sudo -u bitcoin node /usr/bin/spark-wallet` or `sudo -u bitcoin node /home/bitcoin/.npm-global/bin/spark-wallet`?

**Comment by gabridome** [comment]
Yes. I'll report ASAP.

**Comment by gabridome** [comment]
they both  work.

**Comment by gabridome** [comment]
Unfortunately not.
What I can say is that it may be related to node.js installation as root I
am not able to clean.
I have tried but I don't really know if my setup is still "dirty".

Il giorno mar 15 gen 2019, 13:29 darthginger <notifications@github.com> ha
scritto:

**Comment by shesek** [comment]
I think this might've been related to `MemoryDenyWriteExecute=true` and was fixed in https://github.com/shesek/spark-wallet/commit/c46f5f11612a11ffef3732addb4c620330d99cb9. Can you try using the updated `spark-wallet.service` and see if that works?

**Comment by gabridome** [comment]
Thank you. I'll report.

**Comment by darthginger** [comment]
It works indeed.

**Comment by gabridome** [comment]
I may try to rebuild the systemctl startup file.

**Comment by shesek** [comment]
@gabridome Yarg. :( It looks like this might be related to one of the other security hardening measures [defined here](https://github.com/shesek/spark-wallet/blob/c46f5f11612a11ffef3732addb4c620330d99cb9/scripts/spark-wallet.service#L22-L26). Can you try taking all of them off and see what happens?

**Comment by gabridome** [comment]
I will do of course. On a side note I have noticed this after rebuilding the service:
```
sudo systemctl start spark-wallet
Failed to start spark-wallet.service: Unit lightningd.target not found.
```
Should I do something before?

**Comment by shesek** [comment]
> It was not in the built standard version so, I guess I copied that from bitcoind.service

It used to be in spark-wallet's service file, you might've gotten it from there :)

> Failed to start spark-wallet.service: Unit lightningd.target not found.

Yes, it assumes you also have [lightningd configured as a systemd service](https://github.com/ElementsProject/lightning/blob/master/contrib/init/lightningd.service).

If you don't, you can change `Requires` and `After` to `network.target` instead of `lightningd.service`, which will make Spark start immediately after networking is available.

**Comment by k3tan25** [comment]
Starting Spark immediately after networking is available doesn't work for me. I tried `requires` and `after` set to `lightningd.service` as well, but that didn't work either. I think by the time bitcoind and lightningd becomes available, spark-wallet has already fired up and can't connect, spitting out an error on startup. Thoughts?

**Comment by shesek** [comment]
> I tried `requires` and `after` set to `lightningd.service` as well, but that didn't work either. I think by the time bitcoind and lightningd becomes available, spark-wallet has already fired up and can't connect, spitting out an error on startup.

Can you check Spark's logs to see if that's really what's happening? I have it setup on multiple machines with `requires`/`after` set to `lightningd.service` and it does appear to work for me.

You can tell Spark to skip testing the connection on startup with `--no-test-conn` as an CLI argument or `NO_TEST_CONN=1` as an env var. This will delay the connection until the first time a client makes a request to the server. But its better to get it to started after lightningd if possible.

**Comment by k3tan25** [comment]
My bad, I'm an idiot. It works. Apologies. Thanks for looking into it. Any reason the script has it as `lightningd.target` and not `lightningd.service`?

**Comment by shesek** [comment]
Oh, you're absolutely right, dunno how that slipped in! Fixed in c166aeb.
